### PR TITLE
CrayPE update for libunwind

### PR DESCRIPTION
**Pull request by lukebroskop**
Allow for multiple symbol defs at link time (with CCE)

### CODE REVIEW

**Comment by haampie** [comment]
I'm not 100% sure about these `muldef` PRs, seems it's more treating the symptom than actually fixing the problem in the package. Isn't cce clang based? Does clang not have these issues? Or is it more of a problem with the compiler instead of the package? Or is it the cray compiler wrapper adding redundant `-lsomelib` linker flags and should we add the `--cray-bypass-pkgconfig` flag as spack maybe knows better what flags it has to add?

Edit: this is a linker flag, so i'm curious if adding `--cray-bypass-pkgconfig` changes anything?

**Comment by lukebroskop** [comment]
`--cray-bypass-pkgconfig` did not help. 

With CCE/11.0.4, libunwind@1.5.0 does not encounter the multiple definitions problem as seen in libunwind@1.4.0 -- so an update to the libunwind codebase has fixed the issue. Since the current E4S spack.yaml from [github](https://github.com/E4S-Project/e4s/blob/9369465b5277a74952f68534e92cf661fda3658e/spack.yaml#L73) specifies libunwind@1.4.0 I added another conditional to the libunwind package.py for 1.4.0

**Comment by haampie** [comment]
Hi @lukebroskop, I can reproduce the issue in a chroot environment with an "unpolluted" Cray PE setup.

The issue is clang 11's new default:

> -fno-common has been enabled as the default for all targets. Therefore, C code that uses tentative definitions as definitions of a variable in multiple translation units will trigger multiple-definition linker errors. Generally, this occurs when the use of the extern keyword is neglected in the declaration of a variable in a header file. In some cases, no specific translation unit provides a definition of the variable. The previous behavior can be restored by specifying -fcommon.

The same issue was already fixed for GCC 10 in this package by adding `-fcommon`, and the issue was fixed upstream in 1.5.0.

- Can't you just upgrade to 1.5.0 in the E4S spack.yaml?
- Can you fix the issue for `@:1.4.0` for both clang and cce by either adding a `conflict` for them or enabling `-fcommon` too for `clang@11:` and `cc@11:`

Also, can you maybe check whether the other packages `kvtree` and `mercury` with `muldef` fixes would actually just need the `-fcommon` instead for some package versions for both `cce@11:` and `clang@11:`? (Not necessarily in this pr)

File: var/spack/repos/builtin/packages/libunwind/package.py [hunk 1]
```diff
@@ -59,6 +59,10 @@ def flag_handler(self, name, flags):
             if '+pic' in self.spec:
                 wrapper_flags.append(self.compiler.cc_pic_flag)
 
+        elif name == 'ldflags':
+            if self.spec.satisfies('%cce') and self.spec.satisfies('@1.4.0:'):
```
Comment (haampie) [inline var/spack/repos/builtin/packages/libunwind/package.py]: You want to use `@:1.4` here, which is equivalent to `< 1.5`

**Comment by lukebroskop** [comment]
Good call @haampie , I just went ahead and piggybacked off of the `-fcommon` fix for GCC.

Thanks for the help!

File: var/spack/repos/builtin/packages/libunwind/package.py [hunk 2]
```diff
@@ -53,7 +53,8 @@ def flag_handler(self, name, flags):
 
         if name == 'cflags':
             # https://github.com/libunwind/libunwind/pull/166
-            if self.spec.satisfies('@:1.4 %gcc@10:'):
+            if (self.spec.satisfies('@:1.4 %gcc@10:') or
+                self.spec.satisfies('@:1.4 %cce')):
```
Comment (haampie) [inline var/spack/repos/builtin/packages/libunwind/package.py]:
```suggestion
                self.spec.satisfies('@:1.4 %cce')) or
                self.spec.satisfies('@:1.4 %clang@11:')):
```

**Review by haampie — CHANGES_REQUESTED** [review]
Can you fix it for clang too?

**Review by haampie — APPROVED** [review]

**Pull request merged**